sys/dev/isp/isp.c

```diff
@@ -4662,15 +4858,468 @@ isp_parse_flt_2xxx(ispsoftc_t *isp, uint8_t *flt_data)
+		goto check_valid_image;
+	}
+
+	valid_sec_image = true;
+
+	if (sec_image_status.image_status_mask & 1) {
+		isp_prt(isp, ISP_LOGCONFIG,
+		    "Secondary image is active");
+		active_sec_image = true;
+	}
+
+check_valid_image:
+	if (valid_pri_image && active_pri_image)
+		active_regions->global = ISP27XX_PRIMARY_IMAGE;
+
+	if (valid_sec_image && active_sec_image) {
+		if (!active_regions->global ||
+		    isp_compare_image_generation(isp,
+			&pri_image_status, &sec_image_status) < 0) {
+			active_regions->global = ISP27XX_SECONDARY_IMAGE;
+		}
+	}
+
+	isp_prt(isp, ISP_LOGDEBUG0, "active image %s (%u)",
+	    active_regions->global == ISP27XX_DEFAULT_IMAGE ?
+		"default (boot/fw)" :
+	    active_regions->global == ISP27XX_PRIMARY_IMAGE ?
+		"primary" :
+	    active_regions->global == ISP27XX_SECONDARY_IMAGE ?
+		"secondary" : "invalid",
+	    active_regions->global);
+}
+
+static bool isp_risc_firmware_invalid(ispsoftc_t *isp, uint32_t *dword)
+{
+	return
+	    !(dword[4] | dword[5] | dword[6] | dword[7]) ||
+	    !(~dword[4] | ~dword[5] | ~dword[6] | ~dword[7]);
+}
+
+static int
+isp_load_ram(ispsoftc_t *isp, uint32_t *data, uint32_t risc_addr,
+    uint32_t risc_code_size)
+{
+	mbreg_t mbs;
+	int rval = ISP_SUCCESS;
+
+	MEMORYBARRIER(isp, SYNC_REQUEST, 0, ISP_QUEUE_SIZE(RQUEST_QUEUE_LEN(isp)), -1);
+	MBSINIT(&mbs, MBOX_LOAD_RISC_RAM, MBLOGALL, 0);
+	mbs.param[1] = risc_addr;
+	mbs.param[2] = DMA_WD1(isp->isp_rquest_dma);
+	mbs.param[3] = DMA_WD0(isp->isp_rquest_dma);
+	mbs.param[4] = risc_code_size >> 16;
+	mbs.param[5] = risc_code_size;
+	mbs.param[6] = DMA_WD3(isp->isp_rquest_dma);
+	mbs.param[7] = DMA_WD2(isp->isp_rquest_dma);
+	mbs.param[8] = risc_addr >> 16;
+	isp_prt(isp, ISP_LOGDEBUG0,
+	    "LOAD RISC RAM %u (0x%x) words at load address 0x%x",
+	    risc_code_size, risc_code_size, risc_addr);
+	isp_mboxcmd(isp, &mbs);
+	if (mbs.param[0] != MBOX_COMMAND_COMPLETE) {
+		isp_prt(isp, ISP_LOGERR, "F/W download failed");
+		rval = ISP_FUNCTION_FAILED;
+	}
+
+	return (rval);
+}
+
+static int
+isp_load_risc_flash(ispsoftc_t *isp, uint32_t *srisc_addr, uint32_t faddr)
+{
+	fcparam *fcp = FCPARAM(isp, 0);
+	int rval = ISP_SUCCESS;
+	unsigned int segments, fragment;
+	unsigned long i;
+	unsigned int j;
+	unsigned long dlen;
+	uint32_t *dcode;
+	uint32_t risc_addr, risc_size = 0;
+
+	isp_prt(isp, ISP_LOGDEBUG0,
+	    "FW: Loading firmware from flash (0x%x).", faddr);
+
+	dcode = isp->isp_rquest;
+	isp_read_flash_data(isp, dcode, faddr, 8);
+	if (isp_risc_firmware_invalid(isp, dcode)) {
+		snprintf(fcp->fw_version_flash, sizeof(fcp->fw_version_flash),
+		    "invalid");
+		isp_prt(isp, ISP_LOGERR,
+		    "Unable to verify the integrity of flash firmware image.");
+		isp_prt(isp, ISP_LOGERR,
+		    "Firmware data: 0x%08x 0x%08x 0x%08x 0x%08x.",
+		    dcode[0], dcode[1], dcode[2], dcode[3]);
+		return (ISP_FUNCTION_FAILED);
+	} else {
+		for (i = 0; i < 4; i++)
+			fcp->fw_flashrev[i] =
+				be32toh((uint32_t)dcode[4 + i]);
```

Comment:
Why do you need a type conversion here? It is already uint32_t after `[]`.

Reply:
You're right again. Found two more of those and fixed it here.